src/lib.rs

```diff
@@ -277,532 +135,100 @@ impl TsunamiBuilder {
+            })
+            .try_reduce(Vec::new, |mut provs, prov| {
+                provs.extend(prov);
+                Ok(provs)
+            })?;
 
-                        debug!(log, "setting up {} instance", name; "ip" => &machine.public_ip);
-                        f(&mut sess)
-                            .context(format!("setup procedure for {} machine failed", name))
-                            .map_err(|e| {
-                                error!(log, "setup for {} machine failed", name);
-                                e
-                            })?;
-                        info!(log, "finished setting up {} instance", name; "ip" => &machine.public_ip);
+        Ok(Tsunami(providers, log))
+    }
+}
 
-                        machine.ssh = Some(sess);
-                        Ok(())
+/// When this is dropped, the instances are all terminated automatically.
+///
+/// # Note
+/// See caveats for Azure and Baremetal machines.
+#[must_use]
+pub struct Tsunami<L: Launcher>(Vec<L>, slog::Logger);
+
+impl<L: Launcher> Tsunami<L> {
+    /// Use to access the machines. The `HashMap` of machines is keyed by the friendly names
+    /// assigned by the call to [add](TsunamiBuilder::add).
+    /// The returned `Machine`s will live for the lifetime of self.
+    pub fn get_machines<'l>(&'l self) -> Result<HashMap<String, Machine<'l>>, Error> {
+        self.0
+            .par_iter()
+            .map(|prov| prov.connect_instances())
+            .try_fold(
+                HashMap::new,
+                |mut machs, res: Result<HashMap<String, Machine<'l>>, Error>| {
+                    res.and_then(|ms| {
+                        machs.extend(ms);
+                        Ok(machs)
                     })
-                    .filter_map(Result::err)
-            }));
+                },
+            )
+            .try_reduce(HashMap::new, |mut machs, ms| {
+                machs.extend(ms);
+                Ok(machs)
+            })
+    }
 
-            if errors.is_empty() {
-                // 4. invoke F with Machine descriptors
-                let start = time::Instant::now();
-                info!(log, "quiet before the storm");
-                res = Some(
-                    f(machines)
-                        .context("tsunami main routine failed")
-                        .map_err(|e| {
-                            crit!(log, "main tsunami routine failed");
-                            e
-                        })?,
-                );
-                info!(log, "the power of the tsunami was unleashed"; "duration" => start.elapsed().as_secs());
-            }
-        } else {
-            crit!(
-                log,
-                "only {} out of {} machines were started; aborting",
-                running,
-                expected_num
-            );
-        }
+    pub fn logger(&self) -> &slog::Logger {
+        &self.1
+    }
+}
 
-        debug!(log, "all done");
+// prevent destructuring
```

Comment:
Is there a particularly good reason to do so?

Reply:
It's important to prevent the `Vec<L>` in `Tsunami` from escaping right? Otherwise, it could be dropped while the `Machine<'l>`'s returned by `get_machines` were still around.

Or, maybe the borrow checker would catch this?

Reply:
The borrow checker should catch that.